sonar-plugin-api/src/main/java/org/sonar/api/server/ws/Request.java

```diff
@@ -304,4 +335,217 @@ private static long parseLong(String key, String value) {
+      return checkDefaultValueSupplier(defaultValueSupplier).get();
+    }
+  }
+
+  /**
+   * Implementation of {@link Param} where the param is present.
+   */
+  private static final class GenericParam<T> implements Param<T> {
+    private final T value;
+
+    private GenericParam(T value) {
+      this.value = value;
+    }
+
+    protected static <T> Param<T> present(T value) {
+      return new GenericParam<>(value);
+    }
+
+    /**
+     * Always returns true.
+     */
+    public boolean isPresent() {
+      return true;
+    }
+
+    /**
+     * @return the value of the parameter
+     *
+     * @throws IllegalStateException if param is not present.
+     */
+    @Override
+    @CheckForNull
+    public T getValue() {
+      return value;
+    }
+
+    /**
+     * Always returns value of the parameter.
+     *
+     * @throws NullPointerException As per the inherited contract, {@code defaultValueSupplier} can't be null
+     */
+    @Override
+    @CheckForNull
+    public T or(Supplier<T> defaultValueSupplier) {
+      checkDefaultValueSupplier(defaultValueSupplier);
+      return value;
+    }
+  }
+
+  /**
+   * Extends {@link Param} with convenience methods specific to the type {@link String}.
+   */
+  interface StringParam extends Param<String> {
+    /**
+     * Returns a {@link StringParam} object which methods {@link #getValue()} and {@link #or(Supplier)} will
+     * return {@code null} rather than an empty String when the param is present and its value is an empty String.
+     */
+    StringParam emptyAsNull();
+  }
+
+  /**
+   * Implementation of {@link StringParam} where the param is not present.
+   */
+  private enum AbsentStringParam implements StringParam {
+    INSTANCE;
+
+    protected static StringParam absent() {
+      return INSTANCE;
+    }
+
+    /**
+     * Always returns false.
+     */
+    @Override
+    public boolean isPresent() {
+      return false;
+    }
+
+    /**
+     * Always throws a {@link IllegalStateException}.
+     */
+    @Override
+    public String getValue() {
+      throw createGetValueISE();
+    }
+
+    /**
+     * Always returns the value supplied by {@code defaultValueSupplier}.
+     */
+    @Override
+    public String or(Supplier<String> defaultValueSupplier) {
+      return checkDefaultValueSupplier(defaultValueSupplier).get();
+    }
+
+    /**
+     * Returns itself.
+     */
+    @Override
+    public StringParam emptyAsNull() {
+      return this;
+    }
+  }
+
+  /**
+   * Implementation of {@link StringParam} where the param is present.
+   */
+  private static final class StringParamImpl implements StringParam {
+    @CheckForNull
+    private final String value;
+    private final boolean emptyAsNull;
+
+    private StringParamImpl(@Nullable String value, boolean emptyAsNull) {
+      this.value = value;
+      this.emptyAsNull = emptyAsNull;
+    }
+
+    protected static StringParam present(String value) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Remove this "protected" modifier. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2156)